### ISSUE TITLE
creation of wrappers in 3.13.0b0

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#547]
In GitLab by @ editomas on Jun 16, 2020, 19:41

@dbeltrankyl I'm running an experiment (a2wk) with vertical-horizontal wrapper and in the wrapper the jobs SIM&DA&REDUCE. I'm using two starting dates. 

I had a well built wrapper running (tmp/a2wk_ASThread_15922223455715_9840_700.cmd) but when it get to the final REDUCE jobs, while they are running, it creates the following wrapper ( tmp/a2wk_ASThread_15923015194777_9264_24.cmd) with only 24 jobs (SIM&REDUCE), then the next wrapper (tmp/a2wk_ASThread_15923028956807_576_1.cmd) with only 1 job (DA) and continues like that, creating only small wrappers. They keep running because the dependency are respected.

Is there anything I can do? I'm thinking to set the number of chunks for the exp so that it will do only one wrapper and then I will have to continue the experiment.

thanks for any advice you can provide.

@f-macchia @orioljorba @ abunuel @gmontanepinto @mcastril

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 16, 2020, 20:19

i had the debug queue set for some jobs (REDUCE and DA), thanks @f-macchia.  maybe the source of the problem, even though it is strange that it created the big wrapper in the first place despite of the debug queue settings. 

i better re-run and check again things again, before you loose time. hence i close the issue. sorry about the noise.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 16, 2020, 20:19

closed

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 17, 2020, 10:20

@ editomas , just to let you know that there's a new optional QUEUE parameter for the wrappers. This is used in case you run, let's say, SIM jobs in debug, but want to use another QUEUE when they are wrapped (prace, xlarge...).

It was set in #478

@dbeltrankyl , could you add it more explicitly to the docs?

https://autosubmit.readthedocs.io/en/latest/usage/wrappers.html?highlight=wrappers

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 17, 2020, 13:50

thanks for the info about the new parameter, @mcastril.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 22, 2020, 11:52

@dbeltrankyl I re-open the issue since the problem was not due to have the queue=debug. 

After a wrapper is COMPLETED, the next one is not built as expected. 

as suggested by @f-macchia and @mcastril (autosubmit/#548) I'll test it with MIN_WRAPPED

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 22, 2020, 11:52

reopened

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 23, 2020, 16:35

hi @ editomas 

I did some different configuration with your wrappers and seem that with the min_wrapper it works well with the  3.13.0b version update that will be uploaded tomorrow ( I have issues with module installation permissions, waiting for helpdesk). It is working for you in the actual version?

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 25, 2020, 11:00

@dbeltrankyl the first wrapper is currently running again (after it was cancelled on Tuesday for machine problems: jobs were too slow) with the MIN_WRAPPED parameter. I can tell you in about 23 hours, when it finishes, if the next one is created correctly, becasue I had problems only with the creation of a second wrapper.

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 26, 2020, 18:25

@dbeltrankyl the MIN_WRAPPED parameter did not work in my case. In exp a2wk I have used in `conf/autosubmit_a2wk.conf` :

```
[wrapper]
TYPE = horizontal-vertical
MACHINEFILES = STANDARD
JOBS_IN_WRAPPER = SIM&DA&REDUCE
MIN_WRAPPED = 25
```

The first wrapper is as expected: `a2wk_ASThread_15930146888771_9840_700`, i.e. 700 jobs that enter in the 24h I've set  up as `MAX_WALLCLOCK = 24:00` for MN4.

After the last DA job (`a2wk_20100904_101_DA`) is COMPLETED, autosubmit declares: "Jobs ready for marenostrum4: 24", and creates a second wrapper: `a2wk_ASThread_15931371027778_9264_24` with 24 jobs only (12 SIMs and 12 REDUCEs).

So basically:
- it does not wait for `a2wk_ASThread_15930146888771_9840_700` to be whole completed
- it does not apply the `MIN_WRAPPED` parameter.

I'm using `3.13.0b0`. Please let me know if I made a mistake in the configuration or there is somethng to be fixed. thanks in advance.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 29, 2020, 08:34

Hi ,

> it does not wait for `a2wk_ASThread_15930146888771_9840_700` to be whole completed

I think this is the principal issue, I need to code this in a way that it doesn't deadlock all the wrappers ( that was happening before in some cases, I noticed it last week.) 
I need to code it in a different way because right now it is just a "timer" if the wrapper takes too long to finish it will start to create the next wrapper without having the min_wrapped into account, I thought, during the weekend, a solution to this but is still not implemented. 

Meanwhile, maybe you can create an artificial job with the frequency parameter that depends on the latest jobs of this wrapper? a2wk_ASThread_15930146888771_9840_700 this would create an artificial dependency for synchronizing the threads 

When that job is completed, that could only contain a sleep of 1 second, if all next wrapper jobs depend on this artificial job, it should create the next wrapper correctly.

>  it does not apply the `MIN_WRAPPED` parameter

It seems to be applied but with the value MIN_WRAPPED - 1 will check

**Comment by kinow** [comment]
In GitLab by @ editomas on Jun 29, 2020, 13:17

> It seems to be applied but with the value MIN_WRAPPED - 1 will check

Hi @dbeltrankyl,  I don't think it is applied for (`MIN_WRAPPED - 1`) either. If you see in a2wk there are also some other wrappers e.g. with 1 job only (e.g., `a2wk_ASThread_15931384963950_576_1.cmd`) created just after the `a2wk_ASThread_15931371027778_9264_24.cmd` mentioned above, or e.g. wrapper with 2 jobs (`a2wk_ASThread_15931918295927_1152_2.cmd)`. 

I have also tried to use `PRESUBMISSION = True` hoping that the next wrapper will be created correctly if they are all created together at the start of the run. However, after the 700 job correctly built wrapper (`a2wk_ASThread_15931937366812_9840_700`), it puts on hold 7 wrappers that don't apply the MIN_WRAPPED:  `a2wk_ASThread_15931938526714_2128_7, a2wk_ASThread_15931938695641_780_4, a2wk_ASThread_15931939055623_1540_3, a2wk_ASThread_15931939286306_2304_3, a2wk_ASThread_15931939585755_8_2,a2wk_ASThread_15931939826738_768_1 `

at the end the 700 job wrappers fails and I see this message that autosubmit "abandons" the jobs on hold in HPC:

```
^[[33m[WARNING] Only Held Jobs active,Exiting Autosubmit (TIP: This can happen if suspended or/and Failed jobs are found on the workflow) ^[[0m^[[39m
No more jobs to run.
```

I'm aware that `PRESUBMISSION = True` is still in testing mode, so not sure if this is a separate issue.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 29, 2020, 16:29

> I think this is the principal issue, I need to code this in a way that it doesn't deadlock all the wrappers 

I think that Autosubmit should not create wrappers with less jobs that the Minimum, because this is not what one would expect when setting the parameter. On the other hand, I understand the decision because we want to maximize efficiency and not put workflows in a deadlock.

This indeed may be a problem, because in some cases you would prefer to continue running independent jobs one by one (in order to not lose time), until the MIN_WRAPPER condition fulfills. I am thinking of a classic case with sequential chunks and a vertical wrapper. Here you only care for speed and you maybe not want to launch a big job to the queues to only run a small number of jobs, and you prefer to run them separately.

This is a different case, because a particular combination of jobs from two different stardates are needed, for the wrapper to work as expected. In this case we would prefer to produce a deadlock and lose some time in the hope that it is not too much, and for the sake of coherence.

So I see two different use cases, and maybe the easiest and more robust choice so everyone is happy and we don't introduce heuristics that do not fulfill what users expect, is to branch this parameter in two different ones (or what is the same, to create an extra option).

So, in summary:

* MIN_WRAPPED should always be respected
* In one case, Autosubmit would just submit independent chunks as long as there are not at least MIN_WRAPPED jobs WAITING/READY.
* In the other case, Autosubmit will not submit any job from the sections included in the wrapper, until there are at least MIN_WRAPPED jobs to wrap. This will introduce deadlocks, and will mean that (in this particular case) no SIM, DA or REDUCE will be able to run out of a wrapper.

What do you think?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 29, 2020, 16:30

> I'm aware that `PRESUBMISSION = True` is still in testing mode, so not sure if this is a separate issue. 

Yes it is, and as this workflow still have some aspects to improve, I would recommend to wait till the wrappers are fail-safe or test in less complex scenarios.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 30, 2020, 09:11

> This indeed may be a problem, because in some cases you would prefer to continue running independent jobs one by one (in order to not lose time), until the MIN_WRAPPER condition fulfils. I am thinking of a classic case with sequential chunks and a vertical wrapper. Here you only care for speed and you maybe not want to launch a big job to the queues to only run a small number of jobs, and you prefer to run them separately.

Yes, Indeed this is what causes to me to implement that current behaviour.

> So, in summary:
> 
> * MIN_WRAPPED should always be respected
> * In one case, Autosubmit would just submit independent chunks as long as there are not at least MIN_WRAPPED jobs WAITING/READY.
> * In the other case, Autosubmit will not submit any job from the sections included in the wrapper, until there are at least MIN_WRAPPED jobs to wrap. This will introduce deadlocks, and will mean that (in this particular case) no SIM, DA or REDUCE will be able to run out of a wrapper.
> 
> What do you think?

Okay, I'll try to implement an option called `SizePolicy= strict/flexible `(or some like that, could change the name)

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 30, 2020, 10:20

Thanks @dbeltrankyl . I thought about the name, but the Size indeed will always be respected. The difference will be that the workflow will be sending standalone jobs when the criteria is not met.

* If MIN_WRAPPED > 0 and WAIT4WRAP=TRUE then... sleep until it can make a wrapper

We've got to think about the name anyway

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 30, 2020, 12:43

We also have to account for retries. So in the hybrid wrappers we will run retries separately (without a wrapper) so the coherence is preserved.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 2, 2020, 10:29

I changed min wrapper, it uses child and parents of all inner jobs to determinate if it should be put into the wrapper or not, I don't use sleep in the wrapper because it would cause to freeze the status of others jobs that are currently running, instead, you could use the safety timer of the wrapper to delay the check.

There are two provisional new behaviours and also the name is provisional. 

`POLICY = flexible`; default. => If a wrapper with MIN_WRAPPED can't be created, it will send the job sequentially.

`POLICY = strict. `=> If a wrapper with MIN_WRAPPED can't be created, it will wait until it is possible to make the wrapper, this can cause deadlocks so be careful when using this option. I'm trying to find a nicer way to stop autosubmit if a deadlock is detected.

Also when an inner job fails, it will be retried as a sequential job so it shouldn't affect the next wrappers.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 12, 2020, 10:35

mentioned in issue auto-monarch#301

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 2, 2020, 13:35

closed
